kloppy/infra/serializers/tracking/statsperform.py

```diff
@@ -0,0 +1,271 @@
+import logging
+from typing import Dict, Optional, Union, NamedTuple, IO
+from lxml import objectify
+
+from kloppy.domain import (
+    TrackingDataset,
+    DatasetFlag,
+    AttackingDirection,
+    Frame,
+    Point,
+    Point3D,
+    Team,
+    BallState,
+    Period,
+    Orientation,
+    attacking_direction_from_frame,
+    Metadata,
+    Ground,
+    Player,
+    Provider,
+    PlayerData,
+)
+
+from kloppy.utils import Readable, performance_logging
+
+from .deserializer import TrackingDataDeserializer
+
+logger = logging.getLogger(__name__)
+
+
+class StatsperformInputs(NamedTuple):
+    meta_data: IO[bytes]
+    raw_data: IO[bytes]
+
+
+class StatsperformDeserializer(TrackingDataDeserializer[StatsperformInputs]):
+    def __init__(
+        self,
+        limit: Optional[int] = None,
+        sample_rate: Optional[float] = None,
+        coordinate_system: Optional[Union[str, Provider]] = None,
+        only_alive: Optional[bool] = True,
+    ):
+        super().__init__(limit, sample_rate, coordinate_system)
+        self.only_alive = only_alive
+
+    @property
+    def provider(self) -> Provider:
+        return Provider.STATSPERFORM
+
+    @classmethod
+    def __get_periods(cls, tracking):
+        """Gets the Periods contained in the tracking data."""
+        lines = tracking.decode("ascii").splitlines()
+        period_ids = []
+        frame_ids = []
+
+        for line in lines:
+            time_info = line.split(";")[1].split(",")
+            period_ids.append(time_info[1])
+            frame_ids.append(time_info[0])
+
+        unique_period_ids = list(dict.fromkeys(period_ids).keys())
+        periods = []
+
+        for period_id in unique_period_ids:
+            period_start_index = period_ids.index(period_id)
+            period_end_index = (
+                len(period_ids) - period_ids[::-1].index(period_id) - 1
+            )
+            periods.append(
+                Period(
+                    id=int(period_id),
+                    start_timestamp=int(frame_ids[period_start_index]),
+                    end_timestamp=int(frame_ids[period_end_index]),
+                )
+            )
+
+        return periods
+
+    @classmethod
+    def __get_frame_rate(cls, tracking):
+        """gets the frame rate of the tracking data"""
+        lines = tracking.decode("ascii").splitlines()
+        timestamps = []
+        num_frames = 5
+
+        for i in [0, num_frames]:
```

Comment:
I would use a `tuple` instead of `list` in this situation. I was wondering whether we really need a loop, but I figured that it probably doesn't really matter in terms of readability and performance.

I am wondering though whether the current implementation is sufficiently robust. Is it fair to assume that the data is always sorted in this specific way and that no frames are missing?

Reply:
In the new implementation, I calculate the most occurring frame delta (https://github.com/PySport/kloppy/pull/191/commits/f80f3ef35b99e9b2458c13a266b576a94a8beab5). Let me know, whether you think this is robust enough. 

Par exemple:
frames: [0, 100, 200, 500]
deltas: [100, 100, 300]
Result in a found delta of 100 (n=2).